### ISSUE TITLE
[Feature] Folder / Tag / Group sharing

### ISSUE THREAD

**Issue by Ninos** [marius-wieschollek/passwords#27]
Hey there,
is it possible to share complete folders or tags with a user group? Such feature would be awesome, e.g. for board/marketing/presales/customers groups :-)

PS: Thank you for the great alternative, which is much more integrated into nextcloud.

**Comment by Ninos** [comment]
Oh, seems dublicate of #18. Is there a reason, why you closed this ticket?

**Comment by marius-wieschollek** [comment]
The reason why #18 is closed is that it was just a question.

The feature itself is quite difficult to implement. The way Passwords is built prevents users from accessing any data which they do not own. Sharing works by copying the passwords via a privileged background job. This gets a lot more complicated when the object is shared with a group and even more complicated when it is implicitly shared trough a folder or a tag.

Currently the main focus is fixing bugs, then adding missing single-user functionality and providing better encryption options. After that browser extensions and apps will be updated.
It will be probably 2019 before i have the necessary time for this functionality.

**Comment by Ninos** [comment]
Thanks for feedback. In NC13 the server-side-encryption was chanced so groups sharing is possible without re-encryption. This is possible by using the master-key which lower the security but enhance the usability. I think switching to this new workflow and implementing an optional e2e-encryption (based on nc) should bring enough security and also usability :-)

**Comment by Ninos** [comment]
BTW:
For such feature request you can also ask for donations for faster implementation, see also:
https://github.com/nextcloud/passman/issues/243

**Comment by tezukzai** [comment]
This app is just what I have been looking for! Great work.

I second the above. Sharing password folders or tags with circles from the circle app or by LDAP groups would be ideal. I'd be happy to contribute a donation to speed this up.

**Comment by pafcioooo** [comment]
Maybe it could be to start with something basic, like sharing to group would result in sharing each password individually to each member of the group. Of course it should work on groups made in nextcloud level, not only password app level (so the ldap groups can be used too).

Similar way we could add folder sharing, so when folder is shared individual passwords are shared. 

In both options of course there should be some filed in db to track if the password is group/folder shared.

**Comment by Flore-mozfr** [comment]
I second this, very good idea.
My usecase: Scout group, the leaders have each an account on the nextcloud instance. I organize them in groups (cubs, beavers, scouts, rovers...) so they can share with their co-leaders. And in september, we welcome new leaders, some leaders change units (from scouts to rover, for example), so I rearrange the groups.
It would be great to be able to share the passwords the same way, So I don't need to rearrange passwords sharing, just the nextcloud groups.

BTW: I am talking of 30 to 40 users

**Comment by PhantomPhreak** [comment]
I'd like to have folder sharing feature too. Sharing a set of passwords that belongs to something will be more convenient than sharing passwords one by one.

**Comment by igloo777** [comment]
I agree with @PhantomPhreak, it's very actual if you have a big number of passwords.

**Comment by LucyDemooon** [comment]
yes it would be nice we also need an password manager for team but 20 or more user to share one by one is heavy, folder and group share with higher priority pls.
@marius-wieschollek i will pay/spend money to this feature for higher priority

**Comment by alexanderdd** [comment]
came here to ask for this, too =)

@marius-wieschollek multiple people have offered financial contributions (@LucyDemooon @tezukzai ). Can you tell us how people can donate? 

Also, people over at https://github.com/nextcloud/passman/issues/243 seem to have found some possibility to make it happen..

**Comment by Natureshadow** [comment]
At least two organisations I work with are in need of that, too ;)!

**Comment by mokkin** [comment]
For me the same. Group and folder/tag sharing would be really great!
If you are interested in donations, please provide payment credentials such as IBAN or Paypal.

**Comment by Ninos** [comment]
Would also like to donate something if it speeds up the implementation. Just tell us your price, may we can collect some donations :-)

**Comment by mokkin** [comment]
There is an option to collect money via [Paypal](https://www.paypal.com/en/webapps/mpp/money-pools) or on https://www.bountysource.com/
But I think @marius-wieschollek should start this if he accepts this way.

**Comment by git001** [comment]
I would also like to donate for this feature.

**Comment by clarm060** [comment]
I also Second the ability to Share Folders to groups and/or users. It would allow me to switch my team to this wonderful tool!

**Comment by perphilip** [comment]
This would be a great feature indeed, but already allowing to share with groups (specifically circles) would be a great start. Thanks for you work so far!

**Comment by grickard** [comment]
I also would be willing to donate if it would push the development of this feature. Another future feature that would be nice would be the ability to send password requests to people. My organization often needs credentials from clients and it would be nice if I could push a secure password request to them and have it automatically added into Passwords. I could see possible security concerns with this though.

**Comment by MrMEEE** [comment]
+1

**Comment by simondaigre** [comment]
I would also like to donate for this feature too.

**Comment by Djiock** [comment]
Hello !
With my company (heavy Nextcloud users) we need a password sharing solution and, using this one personally, I told it's awesome, it's just missing this essential feature...
We're also willing to pay to have it, if that can be done withing the next few months :)

**Comment by JulesBalgue** [comment]
Hi, 

We are ready to donate for this feature too.

Hope this feature will comes quickly ! :)

**Comment by nomad-lw** [comment]
+1
Would love to see this get implemented.

**Comment by ndom91** [comment]
+1 critical feature for our organization. Any word if there is any progress on this?

**Comment by neni84** [comment]
+1 I confirm that this functionality is essential for use in a company / work group.

**Comment by RailsViceRoads** [comment]
I'm looking for an nextcloud app to share "weak secrets" like codes for combination locks. For this kind of information it will be very useful to share folders with groups of users.

**Comment by Terrox** [comment]
I would like Folder Sharing too. So any future passwords I add will appear for shared users. It would also need to share sub-folders automatically too since I would have Company->Client->password type of structure.

**Comment by jinjanko** [comment]
guys, I really need this feature. How can I pay, or donate, or how could I be sure if I donate that it will be done? thank you very much. btw. Great app for nextcloud.

**Comment by JohannesHoffmann** [comment]
I would join jinjanko with a donation to see this nice featurerequest coming live!

**Comment by Ninos** [comment]
Same :D

**Comment by waynedpj** [comment]
:+1:

**Comment by jcklpe** [comment]
Badly needed feature 👍

**Comment by anlarin** [comment]
This feature is very necessary. Your competitor passman does not have her. You must be the first to have this feature.

**Comment by waja** [comment]
> This feature is very necessary. Your competitor passman does not have her. You must be the first to have this feature.

sorry @anlarin, but what is wrong with you? This is a volunteer project, calm down!

**Comment by roygeraets** [comment]
I would like and invest in this feature. Love the product! +1

**Comment by htc1977** [comment]
> I would like and invest in this feature. Love the product! +1

me, too!

**Comment by ffuentes** [comment]
:+1:

**Comment by Nils-Thiele** [comment]
Hi, is this feature still being considered or will it be dropped? @marius-wieschollek  It's a great feature for a small teams.
Maybe it's also possible to store the password database in a place where all users of the cloud can access it. :)

**Comment by Djiock** [comment]
> Hi, is this feature still being considered or will it be dropped? @marius-wieschollek It's a great feature for a small teams.
> Maybe it's also possible to store the password database in a place where all users of the cloud can access it. :)

Seems to be considered for Q4 this year if we look at the roadmap : https://git.mdns.eu/nextcloud/passwords/wikis/Project/Roadmap
Can't wait ! :)

**Comment by rwagnon** [comment]
My team would like to work on this feature. @marius-wieschollek do you have a work plan / concept that we could follow?

**Comment by marius-wieschollek** [comment]
I had planned to take a closer look at this after i'm done with the browser extension. I haven't made a plan, just some thoughts on the issue. Sharing is a nightmare of edge-cases and management functionality.

I would like to split this ticket into four new ones since folder sharing, tag sharing and group sharing are different features and Client-Side-Encryption should be part of the package. (Also this ticket is full of "+1" comments)

1. *Encryption Issues*
   - CSE is already there in the app, so the easiest way would be to use the existing CSE but add a new keychain that handles the keys of shared objects. 
   - The big issue how to handle the keys. One key per object would result in a huge keychain. One key per account would mean that once someone shared one password with you, you could also decrypt all other shared objects from that user. (Evil admin issue) One key per share would cause conflicts (Sharing a subfolder or sharing also with a tag). Maybe one key per folder would work. Or maybe you can check which elements already have a shared key and then assign a new key to the rest with a maximum number of objects per key.
   - To get the keys to the other user i would imagine some kind of inbox. If you share something, the necessary keys are encrypted with a password, stored on the server and you have to give the password to the other user so that he can accept the share request and add the keys to his keychain.
   - You probably need different keychains for keys you shared and keys shared with you.

2. *Sharing synchronization*
   - The current system causes synchronization issues and misunderstandings for the users.
The app does not allow any user to access the data of any other user. So there needs to be some kind of synchronization.
   - I would imagine a system, where the shared properties of an object are copied to a log table. When a shared object is requested, that table is checked for new data. If there is new data, a new revision is created, if not the current one is used. 
   - If the check discovers that the password is no longer shared, it is deleted and a 404 is returned by the api.
   - The initial creation and regular synchronization would still be done by a cron job.
   - To keep track of which user has which revision of a shared object, i would probably introduce a less unique id, that is the same for each copy of a shared object or revision. This would also replace the id given to the user trough the api to make it clear that the objects are the same.

3. *Folder sharing*
   - Will it be possible to move a password outside the shared folder?
   - How is this handled if you're the owner, and how if you're not?

4. *Tag sharing*
   - Will it be possible to remove the tag if you're not the owner?

5. *Group sharing*
   - How is adding/removing users from the group handled?

**Comment by alexanderdd** [comment]
Hey, if I can offer my two cents:
In our context it would be most useful to have folder sharing and be able to share that folder with individual nextcloud users or nextcloud groups. All password app behavior should follow the same logic as nextcloud files.

My opinion:

Folder sharing - **should be implemented first**
-   Will it be possible  to move a password outside the shared folder? - In the long run, it would be good to have a feature "copy entry to folder...". But it is not necessary now.
-  How is this handled if you're the owner, and how if you're not? - No difference. If people have access to a password, they can also save it in another way, with or without a copying feature. Later on, it would be useful to implement read/write permissions for folders per user/group, and only owner can edit that.

Tag sharing - **not neccessary ATM**

- Will it be possible to remove the tag if you're not the owner? - No.

Group sharing - **not neccessary ATM** (later, it should work the same way as Group Folder App)
- How is adding/removing users from the group handled? - Same logic like data folders in Nextcloud files. Adding a user gives full access, removing a user deletes the folder. (After all, when an owner of a folder or admin removes someone from a share this has happened for a reason.)
------
Maybe it makes sense to write some **user stories** or ask the community (in the forum, on facebook?) to describe some ways how they would like to use password sharing (including must-have and nice-to-have distinction).

**Comment by rwagnon** [comment]
Thank you.  This gives a lot of initial information.  I'll look at the code a bit to learn more about  your observations.
________________________________

**Comment by marius-wieschollek** [comment]
@alexanderdd 
The issue with moving passwords out of a folder is the data structure for everyone else with access to this objects. All passwords and folders need a parent folder. If that is missing, the object will be missing from the ui. There are the following options to work with that:
a) If you move an object out of the shared folder, it will be removed from the shared objects and only exist in the account of the user which moved it. (Nextcloud handles it like that)
b) If you move an object out of the shared folder, the system will ignore that for everyone else. Do we have to mark the object as "don't move" so it will not randomly be moved back if someone else moves it within the shared folder? (Probably the most user friendly)
c) You can't move an object out of the shared folder. You will receive an error message.
d) Folder structure is not synchronized. Only if the owner moves something out of the shared folder, it will be removed from the shared objects. Any moving action by another user only exists in his account. (Easiest method, just like with favorites)

I think a) will cause a lot of support overhead because people will not expect this behavior and will complain about passwords etc. suddenly missing from their accounts. c) will cause people to request that this option is added. d) sounds easy but will set users up for a disaster. You task your intern with bringing some structure in the shared passwords. He does so in his account and deletes no longer used folders afterwards. Suddenly all the passwords in those folders are deleted too, since they were still in those folders in the owner's account.
I would prefer b) or c).


> Group sharing [...] (later, it should work the same way as Group Folder App)

No it wouldn't. This means sharing an object with a group.
Group folders in the app are different from that since they don't have an owner.

**Comment by joostvkempen** [comment]
For my organisation sharing like with 'group folder app' is the most preferable. It is a volunteer organisation that is organised in small, fixed teams. Every team is a Nextcloud group. So every volunteer has his own personal workspace ánd shared workspaces equal to the teams they participate in. 


With the passwords I would like something similar: most of them are personal, only few are shared by the team or even the whole organisation.  Most of these passwords have no organizational owner, they are linked to the team. So when team members change, the password will be persistent to the team. 

For our organisation it's a problem when files (or in this case password objects) are owned by a single user. When this volunteer quits, the file or password will be gone for the rest of a team and for his successor.

As a workaround I sometimes use a admin-account that will share everything with the organisation and the teams. The admin-account is owned by a board member, next to his personal account. In this case sharing a folder to nextcloud groups would be the best option. And being able to set permissions for edit, delete etc.

This would be our use case. I'm not a developer, so can't help with the hard work, but I'm willing to support with testing or further thinking.

**Comment by rwagnon** [comment]
@marius-wieschollek As the primary developer, I think you should simply make the decision (after people comment.)  I would like to see implementation of small practical steps so we can work to move it forward.

I believe option a) is the only realistic option.  I use other systems (LastPass Enterprise and SolarWinds Passportal) and if someone removes something, it is simply gone.  People can look at the audit history to see what happened.  They should be able to move an entry from one Shared Folder to another if they want to keep it in circulation.

**Comment by rwagnon** [comment]
I've been using the Password Share tool and find that it is a decent short-term mechanism AS-IS.  It isn't fancy, but I can share passwords with others.  We can each organize them as we wish.  With user training, I would even have everyone use a common folder structure to organize things.  The two short-comings I notice are that only the owner can edit it and ownership cannot be transferred.

**Comment by rwagnon** [comment]
@marius-wieschollek I think we'll need a formal spec document with use-cases.  We could create and share the document for comment.  Ultimately, contributors get to work on whatever they need most.

**Comment by nicoh88** [comment]
+1 for sharing a password-folder with all included passwords! ❤️

**Comment by Ninos** [comment]
> 
> 
>     1. _Encryption Issues_
>        
>        * CSE is already there in the app, so the easiest way would be to use the existing CSE but add a new keychain that handles the keys of shared objects.
>        * The big issue how to handle the keys. One key per object would result in a huge keychain. One key per account would mean that once someone shared one password with you, you could also decrypt all other shared objects from that user. (Evil admin issue) One key per share would cause conflicts (Sharing a subfolder or sharing also with a tag). Maybe one key per folder would work. Or maybe you can check which elements already have a shared key and then assign a new key to the rest with a maximum number of objects per key.
>        * To get the keys to the other user i would imagine some kind of inbox. If you share something, the necessary keys are encrypted with a password, stored on the server and you have to give the password to the other user so that he can accept the share request and add the keys to his keychain.
>        * You probably need different keychains for keys you shared and keys shared with you.
> 

Hey @marius-wieschollek ,
thx for feedback. About encryption I would solve it like NC-Core. For each password/entry create an own master cert. After sharing with users, use the users private key to access the master cert. After sharing with a tag, use the tag master cert for access to the master cert and the user cert for access to the tag master cert. After user was kicked from tag group, you can renew the tag master cert with the public certs of the users. Same could be done with folder sharing, group-sharing...

I'm not a pro in encryption, but should be the most elegant way.

**Comment by martinadieb** [comment]
are there any news about the feature? It would really solve a lot of problems for many teams ... thanks in advance

**Comment by pascal-ts** [comment]
I'm also really looking forward for this feature!
imho only folder sharing is necessary atm, once shared with other users, the owner can add/move/delete passwords and the changes will sync for everyone.
this would allow me to use Passwords in my company

**Comment by le-patenteux** [comment]
Not only am I adding my name to the request, but I have to say that it is the final hurtle we need to fix to adopt Nextcloud company-wide!

Keep-on the good work, NextCloud is a great tool and the community makes it even greater!

**Comment by tibring** [comment]
Is estimation for this feature the same as published at Roadmap?
I'm waiting for this feature :)

**Comment by kevinkk525** [comment]
I'm actually eagerly waiting for that feature too since 2 years now.. Would make it so much easier to use passwords in my organization.
So even though I can't contribute with code, I can at least show my support for that feature :)

**Comment by rakurtz** [comment]
This is a killer feature that would separate this app from any other password-manager on the market:

Self hosted and secure in your Nextcloud and fully sufficient for teams! 

By now it‘s „just“ a nice passwords app for individuals. 

Really looking forward to this.

**Comment by majkinetor** [comment]
> By now it‘s „just“ a nice passwords app for individuals.

You could use API to share folders now tho, although its far from user friendly...

**Comment by rakurtz** [comment]
How am i gonna do that? That would be awesome. We are a team of 6 - sharing one folder would be absolutely enough!

**Comment by majkinetor** [comment]
You need to make cron job that will use share api to a) get all passwords b) share them to list of people (or even group but that will require also NextCloud API to get people in the group).

Here is something in Powershell that can make you get started:
https://gist.github.com/majkinetor/67037053246f4637bba50c3e3f5372f4#passwords

**Comment by aqos156** [comment]
@marius-wieschollek Hi, is there any ETA for this feature?

**Comment by marius-wieschollek** [comment]
no

**Comment by pReya** [comment]
@marius-wieschollek Is this a problem/task that could be sped up with money? Would you prioritize this feature, if there was some kind of "bounty" for it? E.g. would 1000€ or maybe 5000€ make a difference? Or is this still something that needs more conceptual work, which can't be easily sped up by throwing money at it?

I think there might be quite a large group of people and even companies, willing to shell out quite some money to get this feature. But first we'd need to know, if money can help here, or not.

**Comment by albjeremias** [comment]
I'm just sharing a folder between groups, and using keeweb to manage a kdbx file!

**Comment by rakurtz** [comment]
> I'm just sharing a folder between groups, and using keeweb to manage a kdbx file!

We do the same. Works fine. It‘s not as convenient when you want to set different permissions on different password categories since you would have to create individual kdbx files to share with different users/groups. In that case switching from a set of passwords to another you would have to re-enter the master password when you open another kdbx file. But for our small team it works perfectly fine since we share all passwords with every member.

**Comment by kscherler-TWX** [comment]
This missing feature is the reason my organization can't use it. 
Give me a reason to get off of dashlane. This should be your #1 priority moving forward (as you can see by the guy offering money above to get it done)

**Comment by majkinetor** [comment]
> This should be your #1 priority moving forward (as you can see by the guy offering money above to get it done)

You people need to learn that money can't buy you everything... You can always fork it and pay to someone else to do it who is willing to do so instead of harassing  the author

**Comment by kscherler-TWX** [comment]
> > This should be your #1 priority moving forward (as you can see by the guy offering money above to get it done)
> 
> You people need to learn that money can't buy you everything... You can always fork it and pay to someone else to do it who is willing to do so instead of harassing the author

Just offering friendly advice. Expressing interest in prioritizing an issue over others isn't harassment. I think this is a great add-on and would love to use it, but can't because of this. But yeah "harrassment" lol.

**Comment by Ninos** [comment]
I aready offered a donation, just need some informations (donation link, how much, ?until date?). Everyone who can code well is welcomed to extend this plugin with such functionality (preferred as MR, preferred plugin owner).

**Comment by NEITG** [comment]
Would also consider a donation to push this one along.

**Comment by albjeremias** [comment]
if there is a bounty > 1K i can look into this.. :)

**Comment by Ninos** [comment]
I'll donate 500 €. If solved pretty good and is used by me (will then definitely happen!), additional 500 €.

**Comment by torchsmith** [comment]
+1

**Comment by DanAtIntegrateIT** [comment]
Hi There, I'd happy offer to pay for this solution and assist with testing.  Reach back out and let me know. And we can discuss as an official project backed by my business.

**Comment by IARI** [comment]
Considering the [Roadmap](https://git.mdns.eu/nextcloud/passwords/wikis/Project/Roadmap) and the last posts in this issue by marius-wieschollek, I think it is safe to assume that the author is working on the issue, well aware of the amount of interest present.

For the sake of reference:
Going back to 2018, there have now been 14 people who have explicitly offered donations so far.
2018: tezukzai, LucyDemooon, alexanderdd, mokkin, Ninos, git001
2019: grickard, simondaigre, Djiock, JulesBalgue, jinjanko, JohannesHoffmann
2021: NEITG, DanAtIntegrateIT

**Comment by kscherler-TWX** [comment]
> Considering the [Roadmap](https://git.mdns.eu/nextcloud/passwords/wikis/Project/Roadmap) and the last posts in this issue by marius-wieschollek, I think it is safe to assume that the author is working on the issue, well aware of the amount of interest present.
> 
> For the sake of reference:
> Going back to 2018, there have now been 14 people who have explicitly offered donations so far.
> 2018: tezukzai, LucyDemooon, alexanderdd, mokkin, Ninos, git001
> 2019: grickard, simondaigre, Djiock, JulesBalgue, jinjanko, JohannesHoffmann
> 2021: NEITG, DanAtIntegrateIT

Yeah apparently I'm a pushy asshole for suggesting it though. LOL.

**Comment by staeglis** [comment]
We are also very interested in this feature.

**Comment by alexanderdd** [comment]
To me it does not look like @marius-wieschollek is currently working on this. Here https://git.mdns.eu/nextcloud/passwords/blob/master/Donate.md it says that he or other devs do not accept bounties for features.

@albjeremias offered to look into this issue if the bounty reaches $1000, which should be easy. So how do we collect the money/pledges? Should I make a gofundme page? (I know this is slightly offtopic, feel free to suggest a different place to discuss this)

Also, it looks like people over here found a way to do this https://github.com/nextcloud/passman/issues/243

**Comment by pReya** [comment]
> To me it does not look like @marius-wieschollek is currently working on this. Here [git.mdns.eu/nextcloud/passwords/blob/master/Donate.md](https://git.mdns.eu/nextcloud/passwords/blob/master/Donate.md) it says that he or other devs do not accept bounties for features.

Where does that assumption come from? He's updated his roadmap 3 weeks ago, and group/folder sharing is listed under "Next". So I'd assume it's pretty far up on his agenda, and he'll be working on it, soon.

**Comment by alexanderdd** [comment]
The assumption comes from the fact that this bug is open for more than three years now. Also, there are other things on the https://git.mdns.eu/nextcloud/passwords/wikis/Project/Roadmap before this. And, this issue is in the section "Next", not in "currently working on". That's why I don't think he is currently working on it.

I don't understand why we should wait (maybe another three years?). We have people who want to pay, we have someone who would do it for money, so we only need to figure out how to handle the money. Please @everyone write your ideas, or point to a different place where we can discuss this if it's too off topic.

**Comment by pavieinv** [comment]
> The reason why #18 is closed is that it was just a question.
> 
> The feature itself is quite difficult to implement. The way Passwords is built prevents users from accessing any data which they do not own. Sharing works by copying the passwords via a privileged background job. This gets a lot more complicated when the object is shared with a group and even more complicated when it is implicitly shared trough a folder or a tag.
> 
> Currently the main focus is fixing bugs, then adding missing single-user functionality and providing better encryption options. After that browser extensions and apps will be updated. It will be probably 2019 before i have the necessary time for this functionality.

If it's hard to implement in the front-end, just make an iteration with jQuery! Man this app is brilliant but the inability to share with multiple accounts is a handicap.

**Comment by pavieinv** [comment]
well, i actually figured out a temporary solution... exporting passwords and importing them into another account !
thanks for this app anyway

**Comment by Ninos** [comment]
> well, i actually figured out a temporary solution... exporting passwords and importing them into another account ! thanks for this app anyway

Very sad to read that 😂😂

Last reminder that I'm still willing to pay, just contact me :-) Most likely the main dev with a donation-link (paypal or something like that), but external devs with at least high security knowledge are also likely welcomed :-)

**Comment by eric-benjamin** [comment]
+1 for sharing passwords
+1 for paying/donating USD,EUR,BTC, a finger?

**Comment by Sirianer** [comment]
+1

**Comment by klemenkobetic** [comment]
Also really interested in this functionality.

What about going through https://bountysource.com/ ?

**Comment by rubensflinco** [comment]
+1

**Comment by LokeYourC3PH** [comment]
+1 on the feature, would really enjoy having some Group share function for collaboration or cases where multiple users need to utilize passwords/data for different services.

**Comment by mercury1337** [comment]
+1 Using passwords for a club and it would be awesome if we could share folders with groups.

**Comment by IARI** [comment]
Thanks for the continued active development on the project.
Considering this issue, It looks as if there has not been any development on the sharing branch.

@marius-wieschollek if could you give us just a very a brief update on your current state of mind regarding this issue, I would be very happy

**Comment by FranziskaSom** [comment]
Hi :)
we are also with you ... where can we donate?

we could use, for example, https://opencollective.com/. would that be something?

see examples for nextcloud support https://opencollective.com/floccus

Best regards to all :)

**Comment by Jean-Jaque** [comment]
Same here. Would be happy to donate as we love this plugin !

**Comment by neon-paco** [comment]
+1

**Comment by LokeYourC3PH** [comment]
I started using and integrating Bitwarden into my service. I'd recommend everyone to do so as well and just drop this, it won't happen.

**Comment by rlKoekie** [comment]
This feature request is a prime example of open source projects :-)
I suspect @marius-wieschollek is more than happy to have this feature, but he probably has plenty of other things to do with his time. However, this does not have to be a showstopper: this is open source software, anybody can contribute!
Yes, yes, not everybody can code, and probably even fewer of us can code in such a way that it is up to standard for such a security-sensitive topic (keep in mind: doing this in the wrong way could cause major security problems). 

For those willing to pay money: check if you know somebody who can take on a project like this, and tell them about the bounties already offered here. The coder-for-hire could then spent a bit of time investigating the project and maybe have a chat with @marius-wieschollek about possible ways of doing this. If that works out smoothly, the coder could then work implementing the feature in an acceptable manner.

TLDR; work with what we have: Marius does not have to do anything for us, and we all could take some form of action if we really wanted to!

@marius-wieschollek : thank you for writing and maintaining this code, I still use it every day!

**Comment by mokkin** [comment]
Thank you for you work @marius-wieschollek and thank you for your words @rlKoekie !
This is exactly my opinion too. Sometimes people have to be reminded to the basics while they are demanding something ;)

**Comment by codePau** [comment]
Hi,
I am interested in three different functionalities:
1. Share one password with a group (1 to many)
2. Share several passwords with a person (many to 1)
3. Share several passwords with a group (many to many)

Functionality 1 seemed to be addressed in Issue 311 (https://github.com/marius-wieschollek/passwords/issues/311), which was closed after referencing this conversation (Issue 27). This conversation is about functionality 3, which would be a combination of functionality 1 and 2 so, in my opinion, Issue 311 shouldn't have been closed. 
I am happy to give some traction to this. Would anyone be interested in working in functionality 1 or 2? You can drop me a line at pau@pau.company. @marius-wieschollek it would be great to have a conversation with you. Thanks for all the work put on this app.
Thanks everybody!

**Comment by grey-scaled** [comment]
First off, I'm not a PHP developer. I'm coming from a C++ world. But as most users of the plugin probably don't understand how complex this is, I'd like to paint a picture of how this could work.

First off: Especially the part of sharing with groups seems rather tricky to me. At some point or another, some user will be granting access to data they don't have access to themselves. If not implemented carefully, this will lead to a potentially disastrous security issue!

Second, for sharing groups, we do NOT want to duplicate the actual data. This WILL become a reference nightmare, especially if we allow users to reshare shared passwords. So, what we instead want to do is: keep the passwords and structures the same, and share the encryption keys.

This can be done with a relatively straightforward concept:
- Each user and each group has an own public/private key pair
- Each user has access to all their groups' private keys
- We have an asymmetrically encrypted key storage, and a symmetrically encrypted password storage
- Each password is encrypted with a unique key
- These keys are stored in the key storage, encrypted with a folder key
- To share, just add another entry in the key storage, encrypted with the recipient's public key

Things to consider:
- if a share is revoked, all affected keys must be replaced
- whenever a share is revoked, should we remind the users to change the previously shared passwords?
- how can we handle granting/revoking access if a user who is not a member of a group adds or removes users from a group?
- if a user's account password is reset, they lose all access to their data, or their private key is not stored encrypted

The last two points can be addressed by storing a copy of all users' and/or groups' private keys encrypted with a single master key. This master key should then be only shared with admins (i.e. stored encrypted with their respective public keys), and could be used for changing group memberships and/or recovering users' access to their passwords.

**If I understand the codebase (from roughly skipping through) correctly, this doesn't only affect sharing, but the way how Passwords stores and accesses data as a whole.**

That is why it really takes a lot of effort to implement, and at the same time might make the system a little less secure.

**Comment by ssergio-ll** [comment]
Is there any information on whether there has been progress on this feature? I think this app is excellent, but we don't use it because we can't share folders with groups. In our company, it's essential. It's very cumbersome to share the passwords one by one. For now, we use KeePassX locally and synchronize the file with Nextcloud, but if this functionality is implemented, it would be a great advance that I think would give the app enormous power.

**Comment by battosai30** [comment]
My advice : don't expect this feature ... I switched to VaultWarden.

**Comment by karge-itestra** [comment]
This feature would be very helpful!

**Comment by johnnyq** [comment]
This would be very useful for us too as we have several departments that need to share passwords

**Comment by NikoKS** [comment]
Truly need this feature 🙏

**Comment by Mululu** [comment]
Hello Earth to developers?

The folder / tag / group sharing function has been requested for more than 2 years but nothing has happened?

If you can't do it alone, get some help from other developers.

**Comment by pReya** [comment]
> Hello Earth to developers?
> 
> The folder / tag / group sharing function has been requested for more than 2 years but nothing has happened?
> 
> If you can't do it alone, get some help from other developers.

You should dial back your tone. No one owes you anything. I suppose you never paid anything for this app. So you'd be better off by saying something like "Thank you for this app, which is already pretty good and I've been using for years, without paying anything".

This GitHub repo is not the primary repo for this plugin. If you really want to get answers, you'd be better off here: https://git.mdns.eu/nextcloud/passwords

**Comment by Natureshadow** [comment]
> I suppose you never paid anything for this app.

You should stop assuming things about others. Especially while putting yourself in a position where you tell others how to behave.

**Comment by Mululu** [comment]
If you feel attacked by something like this after 2 years of feature requests, something is wrong. You just want to see an answer or some progress. What's wrong with that!?

I'm happy to pay for software, plug-ins, etc. if they meet the basic requirements. And in my opinion, a sharing function for folders, tags and groups is a must, especially in Nextcloud.

Furthermore, I don't use this app because of the missing feature! As soon as this feature is available, I'll be happy to include it in my Nextcloud.

**Comment by Mululu** [comment]
> > Hello Earth to developers?
> > The folder / tag / group sharing function has been requested for more than 2 years but nothing has happened?
> > If you can't do it alone, get some help from other developers.
> 
> You should dial back your tone. No one owes you anything. I suppose you never paid anything for this app. So you'd be better off by saying something like "Thank you for this app, which is already pretty good and I've been using for years, without paying anything".
> 
> This GitHub repo is not the primary repo for this plugin. If you really want to get answers, you'd be better off here: https://git.mdns.eu/nextcloud/passwords

If you feel attacked by something like this after 2 years of feature requests, something is wrong. You just want to see an answer or some progress. What's wrong with that!?

I'm happy to pay for software, plug-ins, etc. if they meet the basic requirements. And in my opinion, a sharing function for folders, tags and groups is a must, especially in Nextcloud.

Furthermore, I don't use this app because of the missing feature! As soon as this feature is available, I'll be happy to include it in my Nextcloud.

**Comment by simaoafonso-pwt** [comment]
Here we go again...

Every time someone comments on this issue, 66 emails go out. This is not the right place to argue about generic software development sustainability. It's pretty clear this is issue is not being worked on, alternatives have been mentioned, comment only if you have something else to add (like a Pull Request).

@marius-wieschollek Can you close this issue to comments, to avoid all the off-topic noise?

PS: I'm unsubscribing anyway, don't @ me.

**Comment by johnnyq** [comment]
@Mululu we all want the features you want.

you may not be aware and may have thought it was funny but the following comments could and have upset developers and supporters working on this project:

_Hello Earth to developers?_

and

_If you can't do it alone, get some help from other developers._ (This one, FOSS Developers don't have time to just reach out to other developers, developers choose to donate their time to projects like this)

No Worries, Just be careful with your words next time that's all.

**Comment by marius-wieschollek** [comment]
Replaced by #582 & #583. These features will be implemented when anyone who wants them has time to implement them. Until then i recommend you look for a different solution.
